Project 2/k-NNAlgorithms/src/MathFunction.java

```diff
@@ -27,4 +27,176 @@ public static double hammingDistance(ArrayList<String> in1, ArrayList<String> in
+            if(countOfClass>=currMax){
+                maxIndex=i;
+            }
+        }
+        mode=input.get(maxIndex);
+        return mode;
+    }
+
+
+    public static ArrayList<String> processConfusionMatrix(ArrayList<String> results, ArrayList<ArrayList<String>> testData){
+
+        int lengthOfData = results.size();
+        int classIndex = testData.get(0).size()-1;
+
+        //initialize classifications in the confusion matrix as the first row
+        ArrayList<ArrayList<String>> confusionMatrix= new ArrayList<>();
+        confusionMatrix.add(new ArrayList<>());
+        for (int i = 0; i <lengthOfData ; i++) {
+            if(!confusionMatrix.get(0).contains(testData.get(i).get(classIndex))){
+                confusionMatrix.get(0).add(testData.get(i).get(classIndex));
+            }
+            if(!confusionMatrix.get(0).contains(results.get(i))){
+                confusionMatrix.get(0).add(results.get(i));
+            }
+        }
+        //initialize the rest of the matrix as an n by n matrix below the classification row
+        //filled with 0 initially
+        for (int i = 0; i <confusionMatrix.get(0).size(); i++) {
+            confusionMatrix.add(new ArrayList<>());
+            for (int j = 0; j <confusionMatrix.get(0).size() ; j++) {
+                confusionMatrix.get(i+1).add("0");
+            }
+        }
+
+        //populate the matrix by indexing guesses by their
+        //true value and the guess
+        for (int i = 0; i <lengthOfData; i++) {
+            String guess = results.get(i);
+            String actual = testData.get(i).get(testData.get(0).size()-1);
+            int indexOfActual = confusionMatrix.get(0).indexOf(actual);
+            int indexOfGuess = confusionMatrix.get(0).indexOf(guess);
+            int currentValAtPos= Integer.parseInt(confusionMatrix.get(indexOfActual+1).get(indexOfGuess));
+            currentValAtPos++;
+            confusionMatrix.get(indexOfActual+1).set(indexOfGuess, currentValAtPos+"");
+        }
+
+        //calculate truePos, falsePos, falseNeg and totalPos by indexing the confusion
+        //matrix appropriately
+        double truePos=0;
+        double totalPos=0;
+        double falsePos=0;
+        double falseNeg=0;
+        double precisionSum=0;
+        double recallSum=0;
+        for (int i = 1; i <confusionMatrix.size() ; i++) {
+            totalPos+=Integer.parseInt(confusionMatrix.get(i).get(i-1));
+            truePos=Integer.parseInt(confusionMatrix.get(i).get(i-1));
+            for (int j = 0; j <confusionMatrix.get(0).size() ; j++) {
+
+                falseNeg+=Integer.parseInt(confusionMatrix.get(i).get(j));
+                falsePos+=Integer.parseInt(confusionMatrix.get(j+1).get(i-1));
+            }
+
+            falseNeg-=truePos;//get rid of true pos value in these results
+            falsePos-=truePos;
+
+            //calculates the precision and recall for each class
```

Comment:
Can we use block comments for comments that are >2 lines?

Reply:
Yes, I will do that from now on.